### PR TITLE
Read cluster shard count from DB

### PR DESCRIPTION
**Pull request by yux0**
**What changed?**
Read cluster shard count from DB

**Why?**
Read cluster shard count from DB and use in cluster cache.

**How did you test it?**
Unit tests

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

File: temporal/fx.go [hunk 1]
```diff
@@ -733,10 +733,16 @@ func loadClusterInformationFromStore(ctx context.Context, config *config.Config,
 			return err
 		}
 		metadata := item.(*persistence.GetClusterMetadataResponse)
+		shardCount := metadata.HistoryShardCount
+		if shardCount == 0 {
+			// This is to add backward compatibility to the config based cluster connection.
```
Comment (meiliang86) [inline temporal/fx.go:738]: Do I understand correct this only applies to ancient versions of Temporal Server?

Comment (meiliang86) [reply temporal/fx.go:738]: Recent versions should always have this stored in DB?

**Review by meiliang86 — APPROVED** [review]

**Review by yycptt — APPROVED** [review]

**Review by wxing1292 — APPROVED** [review]

File: temporal/fx.go (hunk 1 above)
Comment (yux0) [reply temporal/fx.go:738]: Yes.

Comment (yux0) [reply temporal/fx.go:738]: Yes. If the clusters are using tctl to connect, then it should have the shard count. If the clusters connection is the legacy config-base, the shard count could be 0. We don't want to support the config base connection so I just assume the clusters shard count must be equal if they are using config base connection.

**Pull request merged**